tests/testthat/test-utility-extra.R

```diff
@@ -0,0 +1,400 @@
+    res = m4ma::getLeaders_rcpp(n-1, state_leaders, centres, objects)
+    testthat::expect_equal(res, ref)
+    testthat::expect_equal(res, NULL)
+  }
+)
+
+testthat::test_that(
+  "Get leaders works (when no in-group and only group)", {
+    state_leaders = state
+    group = c(1, 2, 3)
+    names(group) = nms
+    state_leaders$group = group
+    ref = m4ma::getLeaders_r(n, state_leaders, centres,
+                             objects, onlyGroup = TRUE)
+    res = m4ma::getLeaders_rcpp(n-1, state_leaders, centres,
+                                objects, onlyGroup = TRUE)
+    testthat::expect_equal(res, ref)
+    testthat::expect_equal(res, NULL)
+  }
+)
+
+testthat::test_that(
+  "Get leaders works (when no in-group)", {
+    state_leaders = state
+    group = c(1, 2, 3)
+    names(group) = nms
+    state_leaders$group = group
+    ref = m4ma::getLeaders_r(n, state_leaders, centres,
+                             objects)
+    res = m4ma::getLeaders_rcpp(n-1, state_leaders, centres,
+                                objects)
+    testthat::expect_equal(res, ref)
+  }
+)
+
+testthat::test_that(
+  "Get leaders works (when no in-group and no prefer group)", {
+    state_leaders = state
+    group = c(1, 2, 3)
+    names(group) = nms
+    state_leaders$group = group
+    ref = m4ma::getLeaders_r(n, state_leaders, centres,
+                             objects, preferGroup = FALSE)
+    res = m4ma::getLeaders_rcpp(n-1, state_leaders, centres,
+                                objects, preferGroup = FALSE)
+    testthat::expect_equal(res, ref)
+  }
+)
+
+testthat::test_that(
+  "Get leaders works (when different directions)", {
+    state_leaders = state
+    a = c(0, 175, 175)
+    names(a) = nms
+    state_leaders$a = a
+    ref = m4ma::getLeaders_r(n, state_leaders, centres, objects)
+    res = m4ma::getLeaders_rcpp(n-1, state_leaders, centres, objects)
+    testthat::expect_equal(res, ref)
+    testthat::expect_equal(res, NULL)
+  }
+)
+
+testthat::test_that(
+  "Get leaders works (when duplicate leaders)", {
+    state_leaders = state
+    group = c(2, 2, 2)
+    names(group) = nms
+    state_leaders$group = group
+    p2_same = rbind(c(0, 0), c(0.5, 0), c(0, 0.5))
+    rownames(p2_same) = c(nms[1], nms[2], nms[2])
+    state_leaders$p = p2_same
+    ref = m4ma::getLeaders_r(n, state_leaders, centres, objects)
+    res = m4ma::getLeaders_rcpp(n-1, state_leaders, centres, objects)
+    testthat::expect_equal(res, ref)
+  }
+)
+
+testthat::test_that(
+  "Get leaders works (when pick best)", {
+    state_leaders = state
+    group = c(2, 2, 2)
+    names(group) = nms
+    state_leaders$group = group
+    p2_same = rbind(c(0, 0), c(0.5, 0), c(0, 0.5))
+    rownames(p2_same) = nms
+    state_leaders$p = p2_same
+    ref = m4ma::getLeaders_r(n, state_leaders, centres,
+                             objects, pickBest = TRUE)
+    res = m4ma::getLeaders_rcpp(n-1, state_leaders, centres,

```

Comment:
consider the opportunity to add a test where pickBest is FALSE, to test all cases of the function

Reply:
`pickBest = FALSE` is the default for this function, which is why I didn't write a separate test for it.